focus/internals/src/lib/model/selection/project.rs

```diff
@@ -269,21 +318,108 @@ impl TryFrom<&ProjectIndex> for TargetSet {
         Ok(Self {
             optional_projects,
             mandatory_projects,
+            optional_project_set_store,
         })
     }
+
+    pub fn save(&mut self) -> Result<()> {
+        self.optional_project_set_store.0.save()
+    }
+
+    pub fn set_project(
+        &mut self,
+        project_name: String,
+        project: Project,
+        maybe_project_file: Option<String>,
+    ) -> Result<(), ProjectCatalogError> {
+        let project_found = self.optional_projects.sources.get(&project_name);
+
+        let new_project_case = project_found.is_none() && maybe_project_file.is_some();
+        let no_file_case = project_found.is_none() && maybe_project_file.is_none();
+        let existing_project_case = project_found.is_some() && maybe_project_file.is_none();
+        let file_matches_case = project_found.is_some()
+            && maybe_project_file.is_some()
+            && *project_found.unwrap() == maybe_project_file.clone().unwrap();
+        let file_conflicts_case = project_found.is_some()
+            && maybe_project_file.is_some()
+            && *project_found.unwrap() != maybe_project_file.clone().unwrap();
+
+        if new_project_case || file_matches_case || existing_project_case {
+            let project_file =
+                maybe_project_file.unwrap_or_else(|| project_found.unwrap().to_string());
+
+            info!(
+                "Saving {}project {} to file {}.",
+                if new_project_case { "new " } else { "" },
+                project_name,
+                project_file
+            );
+            self.optional_projects
+                .sources
+                .insert(project_name.clone(), project_file.clone());
+            self.optional_projects
+                .underlying
+                .insert(project_name.clone(), project.clone());
+            match self
+                .optional_project_set_store
+                .underlying_mut()
+                .get_mut(&project_file)
+            {
+                Some(project_set) => {
+                    project_set.set_project(project_name, project).unwrap();
+                }
+                None => {
+                    let new_project_set = ProjectSet {
+                        projects: vec![project],
+                    };
+                    self.optional_project_set_store
+                        .0
+                        .insert(project_file.as_str(), &new_project_set)
+                        .unwrap();
+                }
+            }
+            self.optional_project_set_store
+                .save_project_set(&project_file)
+                .unwrap();
+        } else if no_file_case {
+            error!(
+                "Project {} was not found, please specifiy a file to save to.",
+                project_name
+            );
+            return Err(ProjectCatalogError::ProjectExistsElsewhere);
+        } else if file_conflicts_case {
```

Comment:
Unsure when this would be the case. Doesn't this feature allow you to update existing projects?

Reply:
This would be the case when someone wants to update project xyz and save it in file xyz.projects.json, but that project already exists in conflict_file.projects.json